File: password.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Getopt::Long;

my $complexity;
my $length;
my $repeat;

my $options =
{
    'complexity=s' => \$complexity,
    'length=s'     => \$length,
    'repeat=s'     => \$repeat,
};

exit if ! GetOptions($options);

my $alphabets =
[
    [ 0..9 ],
    [ 0..9, 'a'..'f' ],
    [ 0..9, 'a'..'z' ],
    [ 0..9, 'a'..'z', 'A'..'Z' ],
    [ 0..9, 'a'..'z', 'A'..'Z', qw( _ - ) ],
    [ 0..9, 'a'..'z', 'A'..'Z',
           qw( ! " $ % & ' \( \) * + - . / @ [ \\ ] ^ _ ` { | } ~ ), '#', ',' ],
];

if ( defined $complexity && ( $complexity !~ /^[0-9]+$/ || $complexity > $#$alphabets ) )
{
    print "Invalid complexity: $complexity\n";
    exit;
}

if ( defined $length && ( $length !~ /^[0-9]+$/ || $length == 0 ) )
{
    print "Invalid length: $length\n";
    exit;
}

if ( defined $repeat && ( $repeat !~ /^[0-9]+$/ || $repeat == 0 ) )
{
    print "Invalid repeat: $repeat\n";
    exit;
}

$complexity = 3  if ! defined $complexity;
$length     = 32 if ! defined $length;
$repeat     = 1  if ! defined $repeat;

my $alphabet = $alphabets->[$complexity];
my $permutations = scalar( @$alphabet ) ** $length;
my $per = '';
$per = ' per password' if $repeat > 1;
printf "Permutations%s: %.3e\n", $per, $permutations;
printf "Entropy%s: %.0f bits\n", $per, log($permutations) / log(2);
print "\n";

foreach my $i ( 1 .. $repeat )
{
    my $password;

    for ( 1 .. $length )
    {
        $password .= $alphabet->[ int rand $#$alphabet ];
    }

    if ( $repeat != 1 )
    {
        printf '%' . length($repeat) . 's  ', $i;
    }

    print "$password\n";
}

File: password.py
#!/usr/bin/env python3

import argparse
import datetime
import math
import os
import random
import sys

from decimal import Decimal

def c_range(a, b):
    return [chr(c) for c in range(ord(a), ord(b) + 1)]

def natural(value):
    if int(value) < 1:
        raise argparse.ArgumentTypeError('invalid choice: {} (choose from 1, 2, 3, ...)'.format(value))
    return int(value)

alphabets = {
    0: c_range('0', '9'),
    1: c_range('0', '9') + c_range('a', 'f'),
    2: c_range('0', '9') + c_range('a', 'z'),
    3: c_range('0', '9') + c_range('a', 'z') + c_range('A', 'Z'),
    4: c_range('0', '9') + c_range('a', 'z') + c_range('A', 'Z') +
        [ '_', '-' ],
    5: c_range('0', '9') + c_range('a', 'z') + c_range('A', 'Z') +
        [ '!', '"', '$', '%', '&', "'", '(', ')', '*', '+', '-', '.', '/',
          '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~', ')', '#' ]
}

parser = argparse.ArgumentParser(allow_abbrev=False, description='Generate random passwords.')
parser.add_argument('-c', '--complexity',
    type=int, choices=alphabets, default=3, action='store', help='password complexity')
parser.add_argument('-l', '--length',
    type=natural, default=32, action='store', help='password length')
parser.add_argument('-r', '--repeat',
    type=natural, default=1, action='store', help='number of passwords')
parser.add_argument('-w', '--write',
    type=str, action='store', help='write to file (use with care)')
args = parser.parse_args()

alphabet = alphabets[args.complexity]
permutations = len(alphabets[args.complexity]) ** args.length
per = ''
if args.repeat > 1:
    per = ' per password'
print('Permutations' + per + ': {:.3e}'.format(Decimal(permutations)), file=sys.stderr)
print('Entropy' + per + ': {:.0f} bits'.format(math.log(permutations) / math.log(2)), file=sys.stderr)
print(file=sys.stderr)

for i in range(0, args.repeat):
    if args.repeat != 1:
        print('{0:>{1}} '.format(i + 1, len(str(args.repeat))), end='')
    password = ''.join(random.choices(alphabet, k=args.length))
    print(password)

    if args.write is not None:
        with open(args.write, 'a') as f:
            print('[{}] {}'.format(datetime.datetime.now().isoformat(), password), file=f)
